include/exadg/solvers_and_preconditioners/multigrid/multigrid_preconditioner_base.cpp

```diff
@@ -332,39 +428,115 @@ create_geometric_coarsening_sequence(Triangulation<dim, spacedim> const & fine_t
 
     Assert(fine_triangulation, ExcNotImplemented());
 
-    // clone distributed triangulation for coarsening
-    parallel::distributed::Triangulation<dim, spacedim> temp_tria(
+    // clone distributed triangulation and start coarsening
+    parallel::distributed::Triangulation<dim, spacedim> tria_copy(
       fine_triangulation->get_communicator(), fine_triangulation->get_mesh_smoothing());
 
-    temp_tria.copy_triangulation(*fine_triangulation);
-
-    // create coarse meshes
-    for(unsigned int l = (fine_triangulation->n_global_levels() - 1); l > 0; --l)
+    tria_copy.copy_triangulation(*fine_triangulation);
+    tria_copy.coarsen_global();
+
+    MPI_Comm mpi_comm = fine_triangulation->get_communicator();
+
+    // as long as we have enough cells per process, we can perform regular
+    // coarsening with all MPI processes TODO: The number of 400 cells per MPI
+    // process (or 50 if the next refinement were done in 3D) was found to be a
+    // good tradeoff between communication cost and workload size of linear
+    // polynomials, resulting in small run times in preliminary studies. This
+    // could be generalized by a parameter to set in the application files.
+    unsigned int n_cells_per_process = 400;
+    for(unsigned int level = fine_triangulation->n_global_levels() - 1;
+        level > 0 && tria_copy.n_global_active_cells() / Utilities::MPI::n_mpi_processes(mpi_comm) >
+                       n_cells_per_process;
+        --level)
```

Comment:
```suggestion
for(int level = fine_triangulation->n_global_levels() - 2; level >= 0 && ...; --level)
```
Note that the type of the variable then has to be `int` instead of `unsigned int`.